packaging/metadata.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import enum
+import typing
+
+from . import specifiers
+from . import utils
+
+if typing.TYPE_CHECKING:
+    from collections.abc import Iterable
+    from typing import List, Optional, Set, Tuple
+
+    from . import requirements
+    from . import version as packaging_version  # Alt name avoids shadowing.
+
+    # Type aliases.
+    _NameAndEmail = Tuple[Optional[str], str]
+    _LabelAndURL = Tuple[str, str]
+
+
+class InvalidMetadata(ValueError):
+    """
+    Invalid metadata found.
+    """
+
+
+@enum.unique
+class MetadataVersion(enum.Enum):
+
+    """
+    Core metadata versions.
+    """
+
+    V1_0 = "1.0"
+    V1_1 = "1.1"
+    V1_2 = "1.2"
+    V2_1 = "2.1"
+    V2_2 = "2.2"
+    v2_3 = "2.3"
+
+
+@enum.unique
+class DynamicField(enum.Enum):
+
+    """
+    Field names for the `dynamic` field.
+
+    All values are lower-cased for easy comparison.
+    """
+
+    # `Name`, `Version`, and `Metadata-Version` are invalid in `Dynamic`.
+    # 1.0
+    PLATFORM = "platform"
+    SUMMARY = "summary"
+    DESCRIPTION = "description"
+    KEYWORDS = "keywords"
+    HOME_PAGE = "home-page"
+    AUTHOR = "author"
+    AUTHOR_EMAIL = "author-email"
+    LICENSE = "license"
+    # 1.1
+    SUPPORTED_PLATFORM = "supported-platform"
+    DOWNLOAD_URL = "download-url"
+    CLASSIFIER = "classifier"
+    # 1.2
+    MAINTAINER = "maintainer"
+    MAINTAINER_EMAIL = "maintainer-email"
+    REQUIRES_DIST = "requires-dist"
+    REQUIRES_PYTHON = "requires-python"
+    REQUIRES_EXTERNAL = "requires-external"
+    PROJECT_URL = "project-url"
+    PROVIDES_DIST = "provides-dist"
+    OBSOLETES_DIST = "obsoletes-dist"
+    # 2.1
+    DESCRIPTION_CONTENT_TYPE = "description-content-type"
+    PROVIDES_EXTRA = "provides-extra"
+
+
+class Metadata:
+
+    """
+    A representation of core metadata.
+    """
+
+    # A property named `display_name` exposes the value.
+    _display_name: str
+    # A property named `canonical_name` exposes the value.
+    _canonical_name: utils.NormalizedName
+    version: packaging_version.Version
+    platforms: Set[str]
+    summary: str
+    description: str
+    keywords: List[str]
+    home_page: str
+    author: str
+    author_emails: List[_NameAndEmail]
+    license: str
+    supported_platforms: Set[str]
+    download_url: str
+    classifiers: Set[str]
+    maintainer: str
+    maintainer_emails: List[_NameAndEmail]
+    requires_dists: Set[requirements.Requirement]
```

Comment:
Do we want to use sets here? Sets don't preserve order, and users may want to preserve order (for example, to group requirements in some logical order).

Only matters if we plan on adding an API for writing metadata out to an external form.

Reply:
There is a "plan on adding an API for writing metadata out to an external form" 🙂 (at least as `bytes`).

I proposed using sets since the core metadata spec doesn't seem to suggest that order is important. Is that something we want to promote people potentially relying on?

Reply:
My use cases are all about reading existing metadata, manipulating it and writing it out. For that use case, having an order-preserving internal form means that diffs between "before" and "after" files are simpler, manual scanning of the file is easier, etc. Also, people *do* seem to write some fields in a logical order (classifiers, for example, are often grouped logically) and the order in the `METADATA` file is respected by, for example, PyPI when writing the output.

So while I wouldn't want people to rely on order for any *functional* purpose, I think there are readability/cosmetic reasons for preserving order.

Reply:
> classifiers, for example, are often grouped logically ... and the order in the METADATA file is respected by, for example, PyPI when writing the output.

Sure, but can't you get a logical ordering by sorting the classifiers as well?



> I think there are readability/cosmetic reasons for preserving order.

My assumption is the encoding methods would keep logical ordering, but based on what you say below is your use-case then I see why we are viewing this class from two different perspectives.